AK/String.cpp

```diff
@@ -108,6 +108,26 @@ ErrorOr<String> String::from_utf16_be(ReadonlyBytes bytes)
     return result;
 }
 
+ErrorOr<String> String::from_utf16_le_with_replacement_character(ReadonlyBytes bytes)
```

Comment:
Hmm I think the name here might be a bit confusing for what this is doing since `String::from_utf8_with_replacement` exists, but if I understand correctly the semantics are a little different.

The existing `with_replacement` function converts to WTF-8 and will not fix up surrogates, whereas these APIs will.

Reply:
from_utf8_with_replacement` currently models this if it helps: https://encoding.spec.whatwg.org/#utf-8-decode-without-bom

Reply:
ah thank you

Reply:
Hmm, no test result actually seems to be changing when doing the scalar string stuff (at least in tree). It is definitely different though, so now I'm wondering about if doing the scalar string thing here is correct or not?

Reply:
to be entirely honest I have no idea what "the scalar string stuff/thing" means

Reply:
What do you think would be a better name for `String::from_utf16_le_with_replacement_character`?

Reply:
so is this all good to merge then?

Reply:
looks good to me, but Tim is much more well versed in this area than I am

Reply:
I think we pass that test because the test first uses `TextEncoder().encode()` to convert the input string to a `Uint8Array`, and that goes through our `to_well_formed_string` AO at the IDL layer. So the string has lonely surrogates replaced before we even get to `TextDecoder().decode()`, which is where `String::from_utf8_with_replacement_character` comes into play.

However, if we don't use `TextEncoder` first:
```html
<script>
  const input = [0xed, 0xa0, 0x80]; // U+D800

  const decoder = new TextDecoder("utf-8");
  const output = decoder.decode(new Uint8Array(input));

  console.log(output.codePointAt(0).toString(16));
</script>
```

We log "d800", whereas Firefox logs "fffd". So it seems our `String::from_utf8_with_replacement_character` is indeed wrong.